### PR TITLE
kv: refactor CompareWithLocality to use enum

### PR DESCRIPTION
**Pull request by wenyihu6**
Prior to this commit, `CompareWithLocality` returned two boolean values
indicating whether two localities were cross-region and cross-zone. However,
this required callers to perform additional cross-comparison of these boolean
values to make meaningful metrics updates.

To simplify this, this commit introduces a new enum type
`LocalityComparisonType`. It provides four locality comparison results: cross
region, same region cross zone, same region same zone, and undefined
(indicating error behavior). This refactoring allow the caller to directly use
the comparison result without additional operations.

In addition, this commit also updates the logic to classify activities between
different regions as cross-regional, regardless of the zone tiers’
configuration. Initially, cross-region but same-zone tiers activities were
flagged as misconfiguration. After some discussion, we have decided that
regions should be non-overlapping. Hence, same zone tiers from different
regions should still be considered as different zones.

In addition, this commit also includes some refactoring of function 
parameters.

Note that this commit does not change any existing functionality.

Part of: https://github.com/cockroachdb/cockroach/issues/103983

Release note: None

### CODE REVIEW

**Review by kvoli — APPROVED** [review]
:lgtm: last commit

> Reviewed 3 of 8 files at r1, 3 of 6 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105267#-:-NYZg8YZ22qYm2GEkq34:b1vpjq0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2588 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105267#-NYZf2AL-lvO_2RZ9WB6:-NYZf2AL-lvO_2RZ9WB7:b9zw959) ([raw file](https://github.com/cockroachdb/cockroach/blob/da864f092e152cee8d651c664162242079b7a7f4/pkg/kv/kvclient/kvcoord/dist_sender.go#L2588)):*
> ```Go
> 	gatewayNodeDesc, err := ds.nodeDescs.GetNodeDescriptor(fromNodeID)
> 	if err != nil {
> 		log.VEventf(ctx, 2, "failed to perform look up for node descriptor %+v", err)
> ```

Why `%+v` and not `%v`?

File: pkg/kv/kvserver/store_snapshot.go
Comment (tbg) [inline pkg/kv/kvserver/store_snapshot.go:1108]: nit: I don't think you need the local here, right? You can inline this into `updateCrossLocalityMetricsOnSnapshotRcvd`.

**Review by tbg — CHANGES_REQUESTED** [review]
Looking good! Only small changes.

This reviews both commits. There is a lot of churn between the two so that in my opinion, they don't make sense as standalone commits. A better sequence of commits is possible, but there are few benefits to introducing it now.

So my suggestion would be to squash the current two commits with an updated message (and PR description) and closing the old PR (that contains the first commit only).
Please add fix-up commits as new commits (only to be squashed immediately before merge, to save me another full review).

**Comment by wenyihu6** [comment]
> Looking good! Only small changes.
> 
> This reviews both commits. There is a lot of churn between the two so that in my opinion, they don't make sense as standalone commits. A better sequence of commits is possible, but there are few benefits to introducing it now.
> 
> So my suggestion would be to squash the current two commits with an updated message (and PR description) and closing the old PR (that contains the first commit only). Please add fix-up commits as new commits (only to be squashed immediately before merge, to save me another full review).

Ack. Added a new commit "[to be squashed commit](https://github.com/cockroachdb/cockroach/pull/105267/commits/c4eacc9b6d8a605dc5263825cb548e8cedd1e703)" which fixes the comments.

**Comment by tbg** [comment]
Looking good! Now all that's left is to squash all commits into the first one 💪🏽

**Review by tbg — APPROVED** [review]
LGTM once squashed into the first commit.

**Review by tbg — APPROVED** [review]
LGTM after squash, on the assumption that the first commit is identical with what I reviewed in the other PR.

**Comment by wenyihu6** [comment]
Thank you so much for the review, Tobi and Austen! (the last push is for a comment from a separate PR and has been approved.)

bors r=tbg,kvoli

**Pull request merged**
